design-system/src/theme-provider.tsx

```diff
@@ -92,19 +92,47 @@ ThemeProvider.defaultProps = {
   theme: 'default',
 };
 
-const useTheme = (parentSelector = defaultParentSelector): ThemeName => {
+type TUseThemeResult = {
+  currentTheme: ThemeName;
+};
+const useTheme = (parentSelector = defaultParentSelector): TUseThemeResult => {
   const [theme, setTheme] = useState<ThemeName>('default');
   const parentSelectorRef = useRef(parentSelector);
+  const observerRef = useRef(
+    new MutationObserver((mutationList, _observer) => {
+      // Since we are only observing the theme DOM node for changes in its
+      // `data-theme` attribute (configured below when calling `observe` function),
+      // we will receive a single element in the list
+      setTheme(
+        (mutationList[0].target as HTMLElement).dataset.theme as ThemeName
+      );
+    })
+  );
 
   // If we use 'useLayoutEffect' here, we would be trying to read the
   // data attribute before it gets set from the effect in the ThemeProvider
   useEffect(() => {
-    setTheme(
-      (parentSelectorRef.current()?.dataset.theme as ThemeName) || 'default'
-    );
+    const themeDomNode = parentSelectorRef.current();
+    const observer = observerRef.current;
+
+    if (themeDomNode) {
+      // We need to read the current theme after the provider is rendered
+      // to have the actual selected theme (calculated client-side) in the
+      // hook local state
+      setTheme((themeDomNode.dataset.theme as ThemeName) || 'default');
+
+      // We observe the theme DOM node for changes in its `data-theme`
+      // attribute, which is the one we update in the `applyTheme` function
+      observer.observe(themeDomNode, {
+        attributes: true,
+        attributeFilter: ['data-theme'],
+      });
+    }
+
+    return () => observer.disconnect();
   }, []);
 
-  return theme;
+  return { currentTheme: theme };
```

Comment:
Nit: any particular reason for naming this `currentTheme` instead of `theme`?

Reply:
My reasoning is that theme can potentially be changed on runtime so I thought having something like "current" or "selected" in the property name makes it clear the value is for the active theme.

Reply:
Trying to think of similar cases that use this kind of naming convention 🤔

As far as it concerns the hook, given that it's named `useTheme` I would expect the value to be `theme`. If we use `currentTheme` I would expect the hook to be named `useCurrentTheme`.

Anyway, not super important. Up to you to decide.

Reply:
Updated in this commit: 13e39429a2f6cc42a06257d405e6cd5130b12b32